documentation/docs/04-hooks.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Hooks
+---
+
+An optional `src/hooks.js` (or `src/hooks.ts`, or `src/hooks/index.js`) file exports three functions, all optional, that run on the server — **getContext**, **getSession** and **handle**.
+
+> The location of this file can be [configured](#configuration) as `config.kit.files.hooks`
+
+### getContext
+
+This function runs on every incoming request. It generates the `context` object that is available to [endpoint handlers](#routing-endpoints) as `request.context`, and used to derive the [`session`](#hooks-getsession) object available in the browser.
+
+If unimplemented, context is `{}`.
+
+```ts
+type Incoming = {
+	method: string;
+	host: string;
+	headers: Headers;
+	path: string;
+	query: URLSearchParams;
+	body: string | Buffer | ReadOnlyFormData;
+};
+
+type GetContext<Context = any> = {
+	(incoming: Incoming): Context;
+};
+```
+
+```js
+import * as cookie from 'cookie';
+import db from '$lib/db';
+
+/** @type {import('@sveltejs/kit').GetContext} */
+export async function getContext({ headers }) {
+	const cookies = cookie.parse(headers.cookie || '');
+
+	return {
+		user: (await db.get_user(cookies.session_id)) || { guest: true }
+	};
+}
+```
+
+### getSession
+
+This function takes the [`context`](#hooks-getcontext) object and returns a `session` object that is safe to expose to the browser. It runs whenever SvelteKit renders a page.
+
+If unimplemented, session is `{}`.
+
+```ts
+type GetSession<Context = any, Session = any> = {
+	({ context }: { context: Context }): Session;
+};
+```
+
+```js
+/** @type {import('@sveltejs/kit').GetSession} */
+export function getSession({ context }) {
```

Comment:
Not strictly related to this change, but is `getSession` specifically intended to not be asynchronous? If so, is there a technical reason for this, or is it just that we want to encourage people to do everything they need in `getContext` and then they just pare away what they don't want sent to the browser in `getSession`?

Reply:
Hadn't really thought about it, honestly. It does feel a bit asymmetrical, having a sync `getSession` next to an async `getContext`, but I'm not sure there's a convincing use case for async `getSession` and it does feel like a waste of a tick. I guess I don't feel strongly

Reply:
https://github.com/sveltejs/kit/blob/6a46e481d242ac18f4d0112ce53469926e50ff37/packages/kit/src/runtime/server/page.js#L453

This seems like it should accept both async and sync versions, since it's awaiting the result.

Reply:
yeah, past me must have thought it was important for some reason. updated the docs to match the code